proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,363 @@
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/698>`_.
+.. sectnum::
+.. contents::
+
+This proposal proposes extending ``-XQualifiedDo`` to literals, to enable more ergonomic and more powerful syntax than ``OverloadedStrings`` or ``OverloadedLists``.
+
+Motivation
+----------
+
+``OverloadedStrings`` and ``OverloadedLists`` are useful for writing normal string/list syntax that get desugared to non-built-in types. However, they do this via typeclasses, which inherits issues like type inference ambiguity. It's possible these issues are one reason why these extensions, despite being in GHC for a long time, are not defaults in GHC202X language editions. Furthermore, they could provide even more flexibility; one note in the original ``OverloadedLists`` `design <https://gitlab.haskell.org/ghc/ghc/-/wikis/overloaded-lists>`_ mentions the inability to support heterogeneous lists.
+
+These extensions allow extensibility of string and list literals, but there's currently no option to allow extending other literals, like numeric literals. Because the ``fromInteger`` function is a part of ``Num``, there's no way to overload numeric literals without providing a specification for all the ``Num`` operations. Even if we broke out ``fromInteger`` into a separate class, we still inherit the same typeclass issues as overloaded strings/lists.
+
+The initial motivation for this came from `String interpolation <https://github.com/ghc-proposals/ghc-proposals/pull/570>`_, where this syntax could provide a mechanism similar to Javascript's `tagged template literals <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates>`_. But it applies more generally to other literals as well.
+
+Semi-related to https://github.com/ghc-proposals/ghc-proposals/issues/438
+
+Proposed Change Specification
+-----------------------------
+
+High-level Overview
+~~~~~~~~~~~~~~~~~~~
+
+``-XQualifiedLiterals`` enables the following syntaxes for expressions
+
+.. list-table::
+    :align: left
+
+    * - **New expression syntax**
+      - **Desugared expression syntax**
+    * - ``Foo.1``
+      - ``Foo.fromNatural 1``
+    * - ``Foo.(1)``
+      - ``Foo.fromNatural 1``
+    * - ``Foo.(-1)``
+      - ``Foo.fromNegativeInteger (-1)``
+    * - ``Foo.(1.2)``
+      - ``Foo.fromRational 1.2``
+    * - ``Foo."asdf"``
+      - ``Foo.fromString "asdf"``
+    * - ``Foo."""asdf"""``
+      - ``Foo.fromString "asdf"``
+    * - ``Foo.[x, y]``
+      - ``Foo.fromList (x `Foo.listCons` (y `Foo.listCons` Foo.listNil))``
```

Comment:
What if instead of `fromList`, `listCons`, and `listNil`, this mapped to a [`build`](https://hackage.haskell.org/package/base-4.21.0.0/docs/GHC-Base.html#v:build)-like interface? That would address @TeofilC's

> there is no single function we can jump to for a list literal

objection. The heterogenous case is messier but still doable:

```hs
build ::
  (
    forall g.
    (forall b bs. f b -> g bs -> g (b ': bs)) ->
    g '[] ->
      g as
  ) ->
    HList f as
build f = f HCons HNil
```

The `Vector` case, which seems to motivate `fromList`, becomes:

```hs
build :: (forall b. (a -> b -> b) -> b -> b) -> Vector a
build f = V.fromList (GHC.Base.build f)
```

Reply:
Great idea! Updated